docs/configuration.md
# Configuration の仕様

webservで用いる設定ファイルの仕様について述べる｡

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [基本](#%E5%9F%BA%E6%9C%AC)
- [server](#server)
  - [listen](#listen)
  - [server_name](#server_name)
  - [location](#location)
    - [allow_method](#allow_method)
    - [client_max_body_size](#client_max_body_size)
    - [root](#root)
    - [index](#index)
    - [is_cgi](#is_cgi)
    - [cgi_executor](#cgi_executor)
    - [error_page](#error_page)
    - [autoindex](#autoindex)
    - [return](#return)
- [サンプル](#%E3%82%B5%E3%83%B3%E3%83%97%E3%83%AB)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## 基本

ファイルの拡張子は `.conf` である｡

正規表現は正規表現エンジンは作る予定無いので対応してない｡

変数についても対応しない｡

## server

- Required: True
- Multiple: True

Syntax: `server {}`

serverブロック｡これ1つでバーチャルサーバ1つを表す｡

以下の項目が設定可能. (末尾に`{}`がついているものはブロック)

- listen
- server_name
- location{}

また､serverブロックは複数書くことで複数のバーチャルサーバを立てることができる｡

### listen

- Required: True
- Multiple: False

Syntax: `listen <port_number>;` or `listen <address>:<port_number>`

その対象のバーチャルサーバでlistenする｡


### server_name

- Required: False
- Multiple: False

Syntax: `server_name: <host_name> [<host_name>...]`

バーチャルサーバで使用するホスト名を指定する｡

これを設定すると"HTTP Hostヘッダー" を元にどのバーチャルサーバと通信を行うか振り分けられる｡

複数のバーチャルサーバの優先順位は以下のようになっている｡

1. listenディレクティブのアドレスとポートに一致するバーチャルサーバを検索する
1. リクエストのHostヘッダが`server_name`ディレクティブで指定したホスト一致したバーチャルサーバにリクエストを振り分ける
1. どのサーバにも一致しない場合デフォルトサーバにリクエストを振り分ける｡ デフォルトサーバは設定ファイルの一番上に記述したバーチャルサーバが使用される｡

### location

- Required: False
- Multiple: True

Syntax: `location <pattern> {}` or `location_back <pattern> {}`

`location` ディレクティブは前方一致でマッチングを行う｡ `<pattern>` には正規表現は使えず､純粋に前方一致のみ行われることに注意｡

`location_back` ディレクティブは後方一致でマッチングを行う｡ 拡張子でルーティングしたい場合などに使う想定｡ ブロック内で利用できるディレクティブなどは`location`と同じ｡

locationの振り分けの優先順位
1. 前方一致のlocationを判定し､最も長い文字列の`<pattern>`にマッチしたものを選ぶ｡

#### allow_method

- Required: False
- Multiple: False

Syntax: `allow_method <method> [<method>...];`

指定した `<method>` のみを許可する｡ 指定しない場合はすべてのメソッドを拒絶する｡

ディレクティブ引数の `<method>` は1つ以上必要｡

`<method>` は以下の文字列のみ許可する
- `GET`
- `POST`
- `DELETE`

#### client_max_body_size

- Required: False
- Multiple: False

Syntax: `client_max_body_size <size>;`

受信できるリクエストボディの最大サイズをバイト単位で指定する｡

#### root

- Required: True
- Multiple: False

Syntax: `root <path>;`

公開するディレクトリを指定する｡

#### index

- Required: False
- Multiple: False

Syntax: `index <path> [<path>...];`

リクエスト対象がディレクトリだった場合にレスポンスとして返すファイル
`is_cgi on;` の場合は無視される｡

#### is_cgi

- Required: False
- Multiple: False

Syntax: `is_cgi <on_or_off>;`

`is_cgi on;` にするとその対象の`location`ディレクティブに入ってくるリクエストはCGIへのリクエストと解釈される｡

指定しない場合は `is_cgi off;` と同じ扱い｡

#### cgi_executor

- Required: True (if `is_cgi on;`)
- Multiple: False

syntax: `cgi_executor <path_or_file_name>`

`<path_or_file_name>` は `execvpe()` の第一引数に渡される文字列｡

#### error_page

- Required: False
- Multiple: True

Syntax: `error_page <status_code> [<status_code>...] <error_page_path>`

`<status_code>` 時に `<error_page_path>` で指定されたファイルを返す｡

#### autoindex

- Required: False
- Multiple: False

Syntax: `autoindex <on_or_off>;`

`autoindex on;` の場合､リクエスト先がディレクトリだった場合にディレクトリ内ファイル一覧ページを返す｡

指定しない場合は `autoindex off;` と同じ扱い｡

#### return

- Required: False
- Multiple: False

Syntax: `return <path>;`

HTTPステータスコード301と共に`<path>`へリダイレクトする｡

e.g. `return http://localhost/index.html;`

## サンプル

```
server {
  listen 80;
  server_name localhost;

  location / {
    allow_method GET;

    root /var/www/html;
    index index.html index.htm;

    error_page 500 /server_error_page.html;
    error_page 404 403 /not_found.html;
  }

  location /upload {
    allow_method GET POST DELETE;

    client_max_body_dize 1M;

    root /var/www/user_uploads;
    autoindex on;
  }
}

server {
  listen 80;
  server_name www.webserv.com webserv.com;

  location / {
    root /var/www/html;
    index index.html;
  }

  location_back .php {
    is_cgi on;
    cgi_executor php-cgi;
    root /home/nginx/cgi_bins;
  }
}

server {
  listen 8080;
  server_name localhost;

  location / {
    root /var/www/html;
    index index.html;
  }
}

server {
  listen 9090;

  location / {
    return http://localhost:8080/;
  }
}
```


srcs/config/config.cpp
#include "config/config.hpp"

#include <stdint.h>
#include <sys/sysinfo.h>

#include <cassert>
#include <iostream>

#include "config/config_parser.hpp"
#include "http/http_status.hpp"

namespace config {

Config::Config() : servers_() {}

Config::Config(const Config &rhs) {
  *this = rhs;
}

Config &Config::operator=(const Config &rhs) {
  if (this != &rhs) {
    servers_ = rhs.servers_;
  }
  return *this;
}

Config::~Config() {}

bool Config::IsValid() const {
  if (servers_.empty()) {
    return false;
  }
  for (VirtualServerConfVector::const_iterator it = servers_.begin();
       it != servers_.end(); ++it) {
    if (!(it->IsValid())) {
      return false;
    }
  }
  return true;
}

void Config::Print() const {
  for (VirtualServerConfVector::const_iterator it = servers_.begin();
       it != servers_.end(); ++it) {
    it->Print();
    std::cout << "\n";
  }
}

const VirtualServerConf *Config::GetVirtualServerConf(
    const std::string &listen_ip, const PortType listen_port,
    const std::string &server_name) const {
  const VirtualServerConf *virtual_server_conf = NULL;

  for (VirtualServerConfVector::const_iterator it = servers_.begin();
       it != servers_.end(); ++it) {
    if ((listen_ip == kAnyIpAddress || it->GetListenIp() == listen_ip) &&
        it->GetListenPort() == listen_port) {
      if (virtual_server_conf == NULL) {
        virtual_server_conf = &(*it);
      } else if (!virtual_server_conf->IsServerNameIncluded(server_name) &&
                 it->IsServerNameIncluded(server_name)) {
        virtual_server_conf = &(*it);
      }
    }
  }

  return virtual_server_conf;
}

const std::vector<VirtualServerConf> &Config::GetVirtualServerConfs() const {
  return servers_;
}

void Config::AppendVirtualServerConf(
    const VirtualServerConf &virtual_server_conf) {
  servers_.push_back(virtual_server_conf);
}

Config ParseConfig(const std::string &filepath) {
  Parser parser;
  parser.LoadFile(filepath);
  return parser.ParseConfig();
}

// server {
//   listen 8080;
//   server_name localhost;
//
//   location / {
//     allow_method GET;
//
//     root /var/www/html;
//     index index.html index.htm;
//
//     error_page 500 /server_error_page.html;
//     error_page 404 403 /not_found.html;
//   }
//
//   location /upload {
//     allow_method GET POST DELETE;
//
//     client_max_body_dize 1048576;
//
//     root /var/www/user_uploads;
//     autoindex on;
//   }
// }
//
// server {
//   listen 8080;
//   server_name www.webserv.com webserv.com;
//
//   location / {
//     root /var/www/html;
//     index index.html;
//   }
//
//   location_back .php {
//     allow_method GET POST DELETE;
//     is_cgi on;
//     root /home/nginx/cgi_bins;
//   }
// }
//
// server {
//   listen 8888;
//   server_name localhost;
//
//   location / {
//     root /var/www/html;
//     index index.html;
//   }
// }
//
// server {
//   listen 9090;
//
//   location / {
//     return http://localhost:8080/;
//   }
// }
Config CreateSampleConfig() {
  Config config;

  // server 1
  VirtualServerConf vserver1;
  vserver1.SetListenPort("8080");
  vserver1.AppendServerName("localhost");

  LocationConf location_v1_1;
  location_v1_1.SetPathPattern("/");
  location_v1_1.AppendAllowedMethod("GET");
  location_v1_1.SetRootDir("/var/www/html");
  location_v1_1.AppendIndexPages("index.html");
  location_v1_1.AppendIndexPages("index.htm");
  location_v1_1.AppendErrorPages(http::SERVER_ERROR, "/server_error_page.html");
  location_v1_1.AppendErrorPages(http::NOT_FOUND, "/not_found.html");
  location_v1_1.AppendErrorPages(http::FORBIDDEN, "/not_found.html");
  vserver1.AppendLocation(location_v1_1);

  LocationConf location_v1_2;
  location_v1_2.SetPathPattern("/upload");
  location_v1_2.AppendAllowedMethod("GET");
  location_v1_2.AppendAllowedMethod("POST");
  location_v1_2.AppendAllowedMethod("DELETE");
  location_v1_2.SetRootDir("/var/www/user_uploads");
  location_v1_2.SetClientMaxBodySize(1048576);
  location_v1_2.SetAutoIndex(true);
  vserver1.AppendLocation(location_v1_2);

  config.AppendVirtualServerConf(vserver1);

  // server 2
  VirtualServerConf vserver2;
  vserver2.SetListenPort("8080");
  vserver2.AppendServerName("www.webserv.com");
  vserver2.AppendServerName("webserv.com");

  LocationConf location_v2_1;
  location_v2_1.SetPathPattern("/");
  location_v2_1.SetRootDir("/var/www/html");
  location_v2_1.AppendIndexPages("index.html");
  vserver2.AppendLocation(location_v2_1);

  LocationConf location_v2_2;
  location_v2_2.SetPathPattern(".php");
  location_v2_2.SetIsBackwardSearch(true);
  location_v2_2.AppendAllowedMethod("GET");
  location_v2_2.AppendAllowedMethod("POST");
  location_v2_2.AppendAllowedMethod("DELETE");
  location_v2_2.SetIsCgi(true);
  location_v2_2.SetRootDir("/home/nginx/cgi_bins");
  vserver2.AppendLocation(location_v2_2);

  config.AppendVirtualServerConf(vserver2);

  // server 3
  VirtualServerConf vserver3;
  vserver3.SetListenPort("8888");
  vserver3.AppendServerName("localhost");

  LocationConf location_v3_1;
  location_v3_1.SetPathPattern("/");
  location_v3_1.SetRootDir("/var/www/html");
  location_v3_1.AppendIndexPages("index.html");
  vserver3.AppendLocation(location_v3_1);

  config.AppendVirtualServerConf(vserver3);

  // server 4
  VirtualServerConf vserver4;
  vserver4.SetListenPort("9090");

  LocationConf location_v4_1;
  location_v4_1.SetRedirectUrl("http://localhost:8080/");
  vserver4.AppendLocation(location_v4_1);

  config.AppendVirtualServerConf(vserver4);

  return config;
}

}  // namespace config
